test/e2e/framework/profile_gatherer.go

```diff
@@ -0,0 +1,147 @@
+			return fmt.Errorf("Failed to create profiles dir: %v", mkdirErr)
+		}
+	} else if err != nil {
+		return fmt.Errorf("Failed to check existence of profiles dir: %v", err)
+	}
+	return nil
+}
+
+func CheckProfileGatheringPrerequisites() error {
+	if !TestContext.AllowGatheringProfiles {
+		return fmt.Errorf("Can't gather profiles as --allow-gathering-profiles is false")
+	}
+	if TestContext.ReportDir == "" {
+		return fmt.Errorf("Can't gather profiles as --report-dir is empty")
+	}
+	if err := CreateProfilesDirectoryIfNeeded(); err != nil {
+		return fmt.Errorf("Failed to ensure profiles dir: %v", err)
+	}
+	return nil
+}
+
+func GatherApiserverCPUProfile(profileBaseName string) error {
+	return GatherApiserverCPUProfileForNSeconds(profileBaseName, DefaultCPUProfileSeconds)
+}
+
+// Note: This function may take a while to execute as it has to SSH to the master,
+// collect the profile for N seconds and then copy the results over. If you are
+// calling this from within a goroutine, you are responsible for waiting for it
+// to finish before the parent goroutine exits.
+func GatherApiserverCPUProfileForNSeconds(profileBaseName string, n int) error {
+	if err := CheckProfileGatheringPrerequisites(); err != nil {
+		return err
+	}
+	// Get the profile data over SSH and write it to a temp file.
+	getCommand := fmt.Sprintf("curl -s localhost:8080/debug/pprof/profile?seconds=%v", n)
+	sshResult, err := SSH(getCommand, GetMasterHost()+":22", TestContext.Provider)
+	if err != nil {
+		return fmt.Errorf("Failed to execute curl command on master through SSH: %v", err)
+	}
+	var tmpfile *os.File
+	tmpfile, err = ioutil.TempFile("", "apiserver-cpu-profile")
+	if err != nil {
+		return fmt.Errorf("Failed to create temp file for profile data: %v", err)
+	}
+	defer os.Remove(tmpfile.Name())
+	if _, err := tmpfile.Write([]byte(sshResult.Stdout)); err != nil {
+		return fmt.Errorf("Failed to write temp file with profile data: %v", err)
+	}
+	if err := tmpfile.Close(); err != nil {
+		return fmt.Errorf("Failed to close temp file: %v", err)
+	}
+	// Create a graph from the profile data and write it to a png file.
+	cmd := exec.Command("go", "tool", "pprof", "-png", tmpfile.Name())
+	outfilePath := path.Join(TestContext.ReportDir, "ApiserverCPUProfile_"+profileBaseName+".png")
+	var outfile *os.File
+	outfile, err = os.Create(outfilePath)
+	if err != nil {
+		return fmt.Errorf("Failed to create file for the profile graph: %v", err)
+	}
+	defer outfile.Close()
+	cmd.Stdout = outfile
+	if err := cmd.Run(); nil != err {
+		return fmt.Errorf("Failed to run 'go tool pprof': %v", err)
+	}
+	return nil
+}
+
+// Note: This function may take a while to execute as it has to SSH to the master,
+// collect the profile and then copy the results over. If you are calling this
+// from within a goroutine, you are responsible for waiting for it to finish before
+// the parent goroutine exits.
+func GatherApiserverMemoryProfile(profileBaseName string) error {
```

Comment:
Can we somehow refactor these two functions to avoid code duplication? The both look really similar.

Reply:
That's true. Done.